### PR TITLE
Add logging details for exceptions

### PR DESCRIPTION
**Pull request by neha-bhargava**
Fixes # 
#3278 

**Changes proposed in this request**
Log request details in case of exception to help with investigation of issues

**Testing**

**Performance impact**

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/Http/HttpManager.cs [hunk 1]
```diff
@@ -145,19 +145,17 @@ private HttpRequestMessage CreateRequestMessage(Uri endpoint, IDictionary<string
...
                 {
-                    logger.Info("The HTTP request was cancelled. ");
+                    logger.Error("The HTTP request was cancelled. ");
                     throw;
                 }
 
                 logger.Error("The HTTP request failed. " + exception.Message);
                 isRetryable = true;
-
                 timeoutException = exception;
```
Comment (SameerK-MSFT) [inline src/client/Microsoft.Identity.Client/Http/HttpManager.cs:154]: nit: extra lines

File: src/client/Microsoft.Identity.Client/Internal/RequestContext.cs [hunk 2]
```diff
@@ -23,6 +23,10 @@ internal class RequestContext
 
         public CancellationToken UserCancellationToken { get; }
 
+        internal string errorMessage { get; set; }
```
Comment (SameerK-MSFT) [inline src/client/Microsoft.Identity.Client/Internal/RequestContext.cs:26]: C# style casing please i.e. ErrorMessage + ErrorMessageWithPii

File: src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs [hunk 3]
```diff
@@ -84,6 +84,7 @@ public async Task<AuthenticationResult> RunAsync(CancellationToken cancellationT
                 {
                     AuthenticationRequestParameters.LogParameters();
                     LogRequestStarted(AuthenticationRequestParameters);
+                    GenerateErrorLogMessage(AuthenticationRequestParameters);
```
Comment (SameerK-MSFT) [inline src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs:87]: Why is this needed?

File: src/client/Microsoft.Identity.Client/Internal/RequestContext.cs [hunk 4]
```diff
@@ -23,6 +23,10 @@ internal class RequestContext
 
         public CancellationToken UserCancellationToken { get; }
```
Comment (SameerK-MSFT) [inline src/client/Microsoft.Identity.Client/Internal/RequestContext.cs:25]: Two members have been added apparently just for logging. What are the other options?

**Review by SameerK-MSFT — CHANGES_REQUESTED** [review]

File: src/client/Microsoft.Identity.Client/Http/HttpManager.cs [hunk 5]
```diff
@@ -145,19 +145,17 @@ private HttpRequestMessage CreateRequestMessage(Uri endpoint, IDictionary<string
             {
                 if (cancellationToken.IsCancellationRequested)
                 {
-                    logger.Info("The HTTP request was cancelled. ");
+                    logger.Error("The HTTP request was cancelled. ");
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Http/HttpManager.cs:148]: Not an error, at most a warning. Very common to cancel the request in AcquireTokenForDeviceCode

File: src/client/Microsoft.Identity.Client/Internal/RequestContext.cs (hunk 4 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Internal/RequestContext.cs]: Changed

File: src/client/Microsoft.Identity.Client/MsalErrorMessage.cs [hunk 6]
```diff
@@ -420,5 +420,9 @@ public static string CertMustHavePrivateKey(string certificateName)
         public const string OboCacheKeyNotInCache = "The token cache does not contain a token with an OBO cache key that matches the longRunningProcessSessionKey passed into ILongRunningWebApi.AcquireTokenInLongRunningProcess method. Call ILongRunningWebApi.InitiateLongRunningProcessInWebApi method with this longRunningProcessSessionKey first or call ILongRunningWebApi.AcquireTokenInLongRunningProcess method with an already used longRunningProcessSessionKey. See https://aka.ms/msal-net-long-running-obo .";
 
         public const string MultiCloudSupportUnavailable = "Multi cloud support unavailable with broker.";
+
+        public const string RequestFailureErrorMessage = "=== Token Acquisition ({0}) failed.";
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/MsalErrorMessage.cs:424]: why the "==="?
is it for formatting?

File: src/client/Microsoft.Identity.Client/Instance/Validation/AdfsAuthorityValidator.cs [hunk 7]
```diff
@@ -36,6 +36,7 @@ public AdfsAuthorityValidator(RequestContext requestContext)
 
                 if (httpResponse.StatusCode != HttpStatusCode.OK)
                 {
+                    _requestContext.Logger.Error($"Authority validation failed due to invalid authority. Authority: {authorityInfo.CanonicalAuthority}");
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/Instance/Validation/AdfsAuthorityValidator.cs:39]:
```suggestion
                    _requestContext.Logger.Error($"Authority validation failed because the configured authority is invalid. Authority: {authorityInfo.CanonicalAuthority}");
```

Comment (trwalke) [reply src/client/Microsoft.Identity.Client/Instance/Validation/AdfsAuthorityValidator.cs:39]: nit

File: src/client/Microsoft.Identity.Client/Instance/Validation/AdfsAuthorityValidator.cs [hunk 8]
```diff
@@ -47,6 +48,7 @@ public AdfsAuthorityValidator(RequestContext requestContext)
                         a => a.Rel.Equals(Constants.DefaultRealm, StringComparison.OrdinalIgnoreCase) &&
                              a.Href.Equals(resource)) == null)
                 {
+                    _requestContext.Logger.Error($"Authority validation failed due to invalid authority. Authority: {authorityInfo.CanonicalAuthority}");
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/Instance/Validation/AdfsAuthorityValidator.cs:51]: nit:
```suggestion
                    _requestContext.Logger.Error($"Authority validation failed because the configured authority is invalid. Authority: {authorityInfo.CanonicalAuthority}");
```

**Review by trwalke — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/MsalErrorMessage.cs (hunk 6 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/MsalErrorMessage.cs:424]: Yes just for formatting. Actually copied the way it is logged in RequestBase.

**Review by SameerK-MSFT — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs [hunk 9]
```diff
@@ -243,6 +243,13 @@ private static void ThrowServerException(HttpResponse response, RequestContext r
 
             if (shouldLogAsError)
             {
+                requestContext.Logger.ErrorPii(
+                    string.Format(MsalErrorMessage.RequestFailureErrorMessagePii,
+                        requestContext.ApiEvent?.ApiIdString,
+                        requestContext.ServiceBundle.Config.Authority.AuthorityInfo.CanonicalAuthority,
+                        requestContext.ServiceBundle.Config.ClientId), 
+                    string.Format(MsalErrorMessage.RequestFailureErrorMessage,
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs:251]: Maybe also add `requestContext.ServiceBundle.Config.Authority.AuthorityInfo.Host` to non-PII message.

File: src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs [hunk 10]
```diff
@@ -243,6 +243,13 @@ private static void ThrowServerException(HttpResponse response, RequestContext r
 
             if (shouldLogAsError)
             {
+                requestContext.Logger.ErrorPii(
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs:246]: This code would log if we get a response from AAD. But in a scenario when there's an exception sending a request, like a request timing out, the exception just bubbles up the top. 
Top of my head maybe add a try/catch/throw, [somewhere here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/8f817842e88dac2988f3914c6f5321190abd6cbf/src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs#L108-L133) and log the same.

**Review by pmaytak — APPROVED** [review]

**Review by pmaytak — CHANGES_REQUESTED** [review]
Need logging when request fails to send.

File: src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs (hunk 10 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs:268]: Added

**Review by pmaytak — APPROVED** [review]
